resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/KubernetesSparkDependencyDownloadInitContainer.scala

```diff
@@ -0,0 +1,112 @@
+import okhttp3.ResponseBody
+import retrofit2.{Call, Callback, Response}
+
+import org.apache.spark.{SecurityManager => SparkSecurityManager, SparkConf, SparkException}
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.kubernetes.CompressionUtils
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+private trait WaitableCallback[T] extends Callback[T] {
+  private val complete = SettableFuture.create[Boolean]
+
+  override final def onFailure(call: Call[T], t: Throwable): Unit = complete.setException(t)
+
+  override final def onResponse(call: Call[T], response: Response[T]): Unit = {
+    require(response.code() >= 200 && response.code() < 300, Option(response.errorBody())
+      .map(_.string())
+      .getOrElse(s"Error executing HTTP request, but error body was not provided."))
+    handleResponse(response.body())
+    complete.set(true)
+  }
+
+  protected def handleResponse(body: T): Unit
+
+  final def waitForCompletion(time: Long, timeUnit: TimeUnit): Unit = {
+    complete.get(time, timeUnit)
+  }
+}
+
+private class DownloadTarGzCallback(downloadDir: File) extends WaitableCallback[ResponseBody] {
+
+  override def handleResponse(responseBody: ResponseBody): Unit = {
+    Utils.tryWithResource(responseBody.byteStream()) { responseStream =>
+      CompressionUtils.unpackTarStreamToDirectory(responseStream, downloadDir)
+    }
+  }
+}
+
+private[spark] class KubernetesSparkDependencyDownloadInitContainer(sparkConf: SparkConf) {
+
+  private val dependencyServerUri = sparkConf.get(RESOURCE_STAGING_SERVER_URI)
+    .getOrElse(throw new SparkException("No dependency server URI was provided."))
+
+  private val downloadJarsSecretLocation = new File(
+    sparkConf.get(INIT_CONTAINER_DOWNLOAD_JARS_SECRET_LOCATION))
+  private val downloadFilesSecretLocation = new File(
+    sparkConf.get(INIT_CONTAINER_DOWNLOAD_FILES_SECRET_LOCATION))
+  require(downloadJarsSecretLocation.isFile, "Application jars download secret provided" +
+    s" at ${downloadJarsSecretLocation.getAbsolutePath} does not exist or is not a file.")
+  require(downloadFilesSecretLocation.isFile, "Application files download secret provided" +
+    s" at ${downloadFilesSecretLocation.getAbsolutePath} does not exist or is not a file.")
+
+  private val jarsDownloadDir = new File(sparkConf.get(DRIVER_LOCAL_JARS_DOWNLOAD_LOCATION))
+  require(jarsDownloadDir.isDirectory, "Application jars download directory provided at" +
+    s" ${jarsDownloadDir.getAbsolutePath} does not exist or is not a directory.")
+
+  private val filesDownloadDir = new File(sparkConf.get(DRIVER_LOCAL_FILES_DOWNLOAD_LOCATION))
+  require(filesDownloadDir.isDirectory, "Application files download directory provided at" +
+    s" ${filesDownloadDir.getAbsolutePath} does not exist or is not a directory.")
+  private val downloadTimeoutMinutes = sparkConf.get(DRIVER_MOUNT_DEPENDENCIES_INIT_TIMEOUT)
+
+  def run(): Unit = {
```

Comment:
put some logging in here.  Checkpoints I'd love to see:

- "Running" (first log line in main method showing when our code starts running)
- download jar/files each started
- download jar/files each finished
- "Finished" (when we're done and return from the main method)